modules/internal/src/main/scala/higherkindness/mu/rpc/internal/serviceImpl.scala

```diff
@@ -50,19 +63,55 @@ object serviceImpl {
       }
     }
     object TypeTypology {
-      def apply(t: Tree): TypeTypology = t match {
-        case tq"Observable[..$tpts]"                        => MonixObservableTpe(t, tpts.headOption)
-        case tq"_root_.monix.reactive.Observable[..$tpts]"  => MonixObservableTpe(t, tpts.headOption)
-        case tq"Stream[${carrier @ _}, ..$tpts]"            => Fs2StreamTpe(t, tpts.headOption)
-        case tq"_root_.fs2.Stream[${carrier @ _}, ..$tpts]" => Fs2StreamTpe(t, tpts.headOption)
-        case tq"Empty.type"                                 => EmptyTpe(t)
-        case tq"${carrier @ _}[..$tpts]"                    => UnaryTpe(t, tpts.headOption)
+
+      private def identOrSelect(tree: Tree, name: String): Boolean = tree match {
+        case Ident(TypeName(`name`)) | Select(_, TypeName(`name`)) => true
+        case _                                                     => false
+      }
+
+      def apply(t: Tree, responseType: Boolean, F: TypeName): TypeTypology = {
+        val unwrappedType: Tree = {
+          if (responseType) {
+            // Check that the type is wrapped in F as it should be, and unwrap it
+            t match {
+              case tq"$f[$tparam]" if f.toString == F.decodedName.toString => tparam
+              case _ =>
+                c.abort(
+                  t.pos,
+                  "Invalid RPC response type. All response types should have the shape F[...], where F[_] is the service's type parameter."
+                )
+            }
+          } else {
+            // Request type is not wrapped in F[...], so return it as-is
+            t
+          }
+        }
+
+        unwrappedType match {
+          case tq"$tpe[$elemType]" if identOrSelect(tpe, "Observable") =>
```

Comment:
Would we need FQN here and for the `Stream` line below?

Reply:
`identOrSelect` will match `Observable`, both qualified and unqualified. So it will match any of:

* `_root_.monix.reactive.Observable`
* `monix.reactive.Observable`
* `reactive.Observable`
* `Observable`

Unfortunately it will also match `com.mylibrary.Observable`, which is not ideal.

I'll add a comment to the code to explain what it's doing, and try rewriting it to be stricter.

Unfortunately we don't have access to any types at this point, so we have to do the best we can with only syntactic analysis.

Reply:
Rewritten and commented in 768a3e476bd6cf1182aee6cdb6cb0cb0b21b4564